docs/src/diplomacy/adder_tutorial.md

```diff
@@ -0,0 +1,322 @@
+      require(ups.forall(up => up.width == ups.head.width))
+      ups.head
+    }
+  )
+  lazy val module = new LazyModuleImp(this) {
+    node.out.head._1 := node.in.head._1 + node.in.tail.head._1
+  }
+
+  override lazy val desiredName = "Adder"
+}
+```
+
+Our `AdderDriver` randomly generates addends each cycle. Note that each node
+passes on two identical `width` parameters to be consumed by the `Adder` and
+`AdderMonitor`.
+
+```scala mdoc
+/** driver (source) */
+class AdderDriver(width: Int)(implicit p: Parameters) extends LazyModule {
+  val nodeA = new AdderDriverNode(Seq(DownwardParam(width), DownwardParam(width)))
+  val nodeB = new AdderDriverNode(Seq(DownwardParam(width), DownwardParam(width)))
+
+  lazy val module = new LazyModuleImp(this) {
+    // generate random addends
+    val randomAddendA = FibonacciLFSR.maxPeriod(nodeA.out.head._2.width)  // notice that we are using negotiated
+    val randomAddendB = FibonacciLFSR.maxPeriod(nodeB.out.head._2.width)  // parameters here
+
+    // drive signals
+    nodeA.out.foreach { case (addend, _) => addend := randomAddendA }
+    nodeB.out.foreach { case (addend, _) => addend := randomAddendB }
+  }
+
+  override lazy val desiredName = "AdderDriver"
+}
+```
+
+Our `AdderMonitor` prints out the results of the `Adder` each cycle and
+signals an error if the `Adder` returns an incorrect result. It has two
+`AdderMonitorNode`s for taking in addends from the `AdderDriver`, and one
+`AdderMonitorNode` to take in the sum from the `Adder`.
+
+```scala mdoc
+/** monitor (sink) */
+class AdderMonitor(width:Int)(implicit p: Parameters) extends LazyModule {
+  val nodeA   = new AdderMonitorNode(UpwardParam(width))
+  val nodeB   = new AdderMonitorNode(UpwardParam(width))
+  val nodeSum = new AdderMonitorNode(UpwardParam(width))
+
+  lazy val module = new LazyModuleImp(this) {
+    val io = IO(new Bundle {
+      val error = Output(Bool())
+    })
+
+    // print output
+    printf(p"${nodeA.in.head._1} + ${nodeB.in.head._1} = ${nodeSum.in.head._1}")
+
+    // basic correctness checking
+    io.error := nodeSum.in.head._1 =/= nodeA.in.head._1 + nodeB.in.head._1
+  }
+
+  override lazy val desiredName = "AdderMonitor"
+}
+```
+
+### Creating the Top
+
+Our top-level module will be a test bench that instantiate our `Adder` and its
+peripherals. Notice that different `width` parameters are passed into Diplomacy
+via `driver` and `checker`. As described in our `AdderNodeImp`, we expect the
+smaller of the two widths to actually be used.
+
+In this module, we also bind our nodes. Sinks are on the left-hand side, while
+sources are on the right.
+
+```scala mdoc
+/** top-level connector */
+class AdderTestHarness()(implicit p: Parameters) extends LazyModule {
+  val adder = LazyModule(new Adder)
+  val driver = LazyModule(new AdderDriver(8))    // passing in downward-traveling width to Diplomacy
+  val checker = LazyModule(new AdderMonitor(4))  // passing in upward-traveling width
+
+  // node connections
```

Comment:
`// Create edges via binding operators between nodes, in order to define a complete graph.`